File: Mathematics/Multiply Strings/leetcode/43. Multiply Strings/Claude/Multiply-Strings.py
# Python版の文字列掛け算アルゴリズムを実装しました。

# **主な特徴：**

# 1. **型安全性の確保:**
#    - すべての変数に明示的な型注釈を追加
#    - `from typing import List` でList型をインポート
#    - Pylanceの型チェックに完全対応

# 2. **パフォーマンス最適化:**
#    - **時間計算量:** O(m × n)
#    - **空間計算量:** O(m + n)
#    - 早期終了による"0"ケースの最適化
#    - リスト内包表記 `[0] * (m + n)` による効率的な初期化
#    - `map(str, result[start:])` による効率的な文字列変換

# 3. **Pythonらしい実装:**
#    - `range(m - 1, -1, -1)` による逆順ループ
#    - `//` 演算子による整数除算
#    - `''.join()` による効率的な文字列結合
#    - プライベートメソッド `_convert_to_string()` による責任分離

# 4. **LeetCode最適化:**
#    - クラス形式 `class Solution` で実装
#    - メモリ使用量を最小限に抑制
#    - 不要な文字列操作を回避
#    - CPython 3.11.4の最適化機能を活用

# 5. **型エラー対策:**
#    - すべての変数に明示的な型注釈
#    - 戻り値の型も明記
#    - Pylanceによる静的解析に完全対応

# この実装はLeetCodeでの実行時間とメモリ使用量を最適化し、Python特有の機能を活用しつつ、型安全性を保証しています。

from typing import List

class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        """
        文字列として表現された2つの非負整数を掛け算する
        
        Args:
            num1 (str): 第1の数値（文字列）
            num2 (str): 第2の数値（文字列）
        
        Returns:
            str: 掛け算の結果（文字列）
        
        Time Complexity: O(m * n) where m = len(num1), n = len(num2)
        Space Complexity: O(m + n) for the result array
        """
        # 特殊ケース：どちらかが"0"の場合、早期終了でパフォーマンス向上
        if num1 == "0" or num2 == "0":
            return "0"
        
        # 文字列の長さを取得
        m: int = len(num1)
        n: int = len(num2)
        
        # 結果を格納する配列（最大桁数は m + n）
        # リスト内包表記を使用してメモリ効率を最適化
        result: List[int] = [0] * (m + n)
        
        # 筆算の掛け算を実装（右端から処理）
        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                # 現在の桁の掛け算
                # int()関数で文字から数値に変換
                mul: int = int(num1[i]) * int(num2[j])
                
                # 結果配列での位置計算
                p1: int = i + j      # 上位桁（繰り上がり先）
                p2: int = i + j + 1  # 下位桁（現在の結果）
                
                # 現在の値に掛け算結果を加算
                sum_val: int = mul + result[p2]
                
                # 下位桁に余りを設定（1桁のみ保持）
                result[p2] = sum_val % 10
                
                # 上位桁に繰り上がりを加算
                result[p1] += sum_val // 10
        
        # 結果配列を文字列に変換（先頭の0を除去）
        return self._convert_to_string(result)
    
    def _convert_to_string(self, result: List[int]) -> str:
        """
        結果配列を文字列に変換し、先頭の0を除去する
        
        Args:
            result (List[int]): 計算結果の配列
        
        Returns:
            str: 文字列形式の結果
        """
        # 先頭の0をスキップする開始位置を見つける
        start: int = 0
        while start < len(result) and result[start] == 0:
            start += 1
        
        # スライスとjoinを使用して効率的な文字列変換
        # map()とstr()を使用してint -> str変換
        return ''.join(map(str, result[start:]))

File: Mathematics/Multiply Strings/leetcode/43. Multiply Strings/Claude/Multiply-Strings.js
// この問題は文字列として表現された2つの大きな数の掛け算を行う問題です。筆算の掛け算アルゴリズムを実装します。この解法の特徴：

// **アルゴリズムの説明：**
// 1. 筆算の掛け算と同じ方法で実装
// 2. 各桁同士の掛け算結果を適切な位置に配置
// 3. 繰り上がりを正しく処理

// **時間計算量：** O(m × n) - num1の長さがm、num2の長さがn
// **空間計算量：** O(m + n) - 結果格納用の配列

// **最適化のポイント：**
// - 特殊ケース（"0"）を最初に処理してパフォーマンス向上
// - 配列のインデックス計算を効率化
// - 不要な文字列操作を避けて、最後に一度だけ文字列変換
// - 先頭の0を効率的に除去

// この実装はLeetCodeの制約（文字列長最大200）に対して十分高速で、メモリ使用量も最小限に抑えています。

/**
 * 文字列として表現された2つの非負整数を掛け算する
 * @param {string} num1 - 第1の数値（文字列）
 * @param {string} num2 - 第2の数値（文字列）
 * @return {string} - 掛け算の結果（文字列）
 */
var multiply = function (num1, num2) {
    // 特殊ケース：どちらかが"0"の場合
    if (num1 === '0' || num2 === '0') {
        return '0';
    }

    const m = num1.length;
    const n = num2.length;

    // 結果を格納する配列（最大桁数は m + n）
    const result = new Array(m + n).fill(0);

    // 筆算の掛け算を実装
    // 右端（最下位桁）から処理
    for (let i = m - 1; i >= 0; i--) {
        for (let j = n - 1; j >= 0; j--) {
            // 現在の桁の掛け算
            const mul = (num1[i] - '0') * (num2[j] - '0');

            // 結果配列での位置
            const p1 = i + j; // 上位桁
            const p2 = i + j + 1; // 下位桁

            // 現在の値に掛け算結果を加算
            const sum = mul + result[p2];

            // 下位桁に余りを設定
            result[p2] = sum % 10;

            // 上位桁に繰り上がりを加算
            result[p1] += Math.floor(sum / 10);
        }
    }

    // 結果配列を文字列に変換（先頭の0を除去）
    let start = 0;
    while (start < result.length && result[start] === 0) {
        start++;
    }

    return result.slice(start).join('');
};
